src/main/scala/cromiam/webservice/CromIamApiService.scala

```diff
@@ -44,6 +53,50 @@ trait CromIamApiService extends Directives with SprayJsonSupport with DefaultJso
     Http().singleRequest(cromwellRequest)
   }
 
+  private[webservice] def authorizeThenForwardToCromwell(user: Option[String], workflowId: String, action: String, request: HttpRequest): Future[HttpResponse] = {
+    (for {
+      _ <- completesOnceUploadComplete(request)
+      _ <- requestAuth(WorkflowAuthorizationRequest(user.getOrElse("anon"), workflowId, action))
+      resp <- forwardToCromwell(request)
+    } yield resp) recover {
+      case SamIamDenialException => HttpResponse(status = Unauthorized, entity = "{}")
+    }
+  }
+
+  private[webservice] def authorizeReadThenForwardToCromwell(user: Option[String], workflowId: String, request: HttpRequest): Future[HttpResponse] =
+    authorizeThenForwardToCromwell(user, workflowId, "read", request)
+  private[webservice] def authorizeAbortThenForwardToCromwell(user: Option[String], workflowId: String, request: HttpRequest): Future[HttpResponse] =
+    authorizeThenForwardToCromwell(user, workflowId, "abort", request)
+
+  private[webservice] def authorizeSubmitThenForwardToCromwell(user: Option[String], request: HttpRequest, batch: Boolean): Future[HttpResponse] = {
+
+    def extractWorkflowIds(response: HttpResponse): Future[List[String]] = {
+      if (batch) {
+        Unmarshal(response).to[List[CromwellStatus]].map(_.map(_.id))
+      } else {
+        Unmarshal(response).to[CromwellStatus].map(s => List(s.id))
+      }
+    }
+
+    val userNonOptional = user.getOrElse("anon")
+    (for {
+      _ <- completesOnceUploadComplete(request)
+      _ <- requestAuth(CreateWorkflowAuthorizationRequest(userNonOptional))
+      resp <- forwardToCromwell(request)
+      workflowIds <- extractWorkflowIds(resp)
+      _ <- registerCreation(userNonOptional, workflowIds)
+    } yield resp) recover {
+      case SamIamDenialException => HttpResponse(status = Unauthorized, entity = "{}")
+    }
+  }
+
+  // To prevent "you replied before the upload finished" errors
+  // TODO: Doesn't work. Waiting for lightbend response.
+  private def completesOnceUploadComplete(r: HttpRequest): Future[Unit] = {
```

Comment:
As I currently understand it (but I have an open lightbend support ticket):

- I'm forwarding the requests through to Cromwell rather than reading the contents and making new requests
- Akka HTTP doesn't load the entire request into memory, it streams them directly through to Cromwell while it's still reading
- If the Cromwell server doesn't pull through every single byte of the stream for whatever reason (I'm not sure why, it's clearly reading the requests), then this doesn't close and my response to the client happens before the client upload stream has been closed. For that, I get a warning message.

So this method forces akka to read the entire request into memory before continuing. Since this solution didn't work I've got a support ticket open, so watch this space for changes.

Most importantly then, unless you're forwarding requests via akka http, I'd say you're probably safe.